src/main/java/frc/robot/subsystems/tank/PoseEstimatorThread.java

```diff
@@ -0,0 +1,116 @@
+package frc.robot.subsystems.tank;
+
+import com.kauailabs.navx.frc.AHRS;
+import com.revrobotics.RelativeEncoder;
+
+import edu.wpi.first.math.MatBuilder;
+import edu.wpi.first.math.Nat;
+import edu.wpi.first.math.estimator.DifferentialDrivePoseEstimator;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.kinematics.DifferentialDriveWheelSpeeds;
+import edu.wpi.first.wpilibj.Timer;
+import edu.wpi.first.wpilibj.Watchdog;
+import edu.wpi.first.wpilibj2.command.WaitCommand;
+
+public class PoseEstimatorThread {
+
+    private final PoseEstimatorRunnable runnable;
+
+    public PoseEstimatorThread(AHRS ahrs, RelativeEncoder leftEncoder, RelativeEncoder rightEncoder) {
+        runnable = new PoseEstimatorRunnable(ahrs, leftEncoder, rightEncoder);
+
+        Thread thread = new Thread(runnable);
+        thread.setDaemon(true);
+        thread.start();
+    }
+
+    public Pose2d getPosition() {
+        return runnable.getPosition();
+    }
+
+    public void setPosition(Pose2d pose) {
+        runnable.setPosition(pose);
+    }
+
+    public DifferentialDriveWheelSpeeds getWheelSpeeds() {
+        return runnable.getLastWheelSpeeds();
+    }
+
+    class PoseEstimatorRunnable implements Runnable {
+        private final DifferentialDrivePoseEstimator poseEstimator;
+
+        private final AHRS ahrs;
+
+        private final RelativeEncoder leftEncoder;
+        private final RelativeEncoder rightEncoder;
+
+        private DifferentialDriveWheelSpeeds lastWheelSpeeds;
```

Comment:
Why is this saved as a field? Is constructing the `DifferentialDriveWheelSpeeds` object really that expensive?

Reply:
No, but calling the encoders again is -- that's what I have been saying about minimizing shuffleboard calls, .set calls, .get calls... I/O takes time

Reply:
In addition to @e3l 's comments:

Because you're essentially running an embedded system, you should think carefully about when objects are created and destroyed. Contrary to typical programming, it's common in embedded to have global variables just so the memory is all allocated at the beginning and there's little to no dynamic memory allocation.

Secondly, any operation like constructing an object has a processing penalty. In the embedded case, it is often better to do this all at startup rather than in your control loop. You want very consistent timing for your control loops, and thus you should prefer to keep the operations in those loops well known.